<a href="https://colab.research.google.com/github/habeebperwad/mlesa/blob/master/Assignment%2003.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np

def J(w):
    return w[0]**2 + w[1]**2 + 4*w[0] - 6*w[1] - 7

def gradientJ(w):
    return np.array([2.0*w[0]+4, 2.0*w[1]-6])

# Do the iteration specified in the iterate-parameter  
# Return the updated w 
# Note: If you want to get J(w) of fifth itertion, use the updated w from fourth iteration. 
def wUpdatedIterate(wCurrent, alpha, iterate):
  
  print("%-2s %-18s %-18s %-11s %-18s" 
      % ("k", "w", "gradient J", "J(w)", "new w"))
  print("-----------------------------------------------------------")

  w = wCurrent
  for k in range(1, iterate+1):
    print("%02d %-18s %-18s %07.4f " 
          % (k, np.round(w,4), np.round(gradientJ(w),4), J(w)), end="")
    w = w - alpha * gradientJ(w)   
    print(" %-18s" % (np.round(w,4)))
      
  return w

#### Questions 1 to 6


- Go to https://www.dcode.fr/minimum-function
- Add **`(x)^2+(y)^2+4*x-6*y-7`** as "FUNCTION (F(X)=)"
-  Add **`x,y`** as 'WITH RESPECT TO'
- click *CALCULATE MINIMUM* button


In [31]:
w50th = wUpdatedIterate(np.array([5, 5]), 0.3, 49) # [5,5] is the value after first itertaion. so need to do only 49 itertaion. 

k  w                  gradient J         J(w)        new w             
-----------------------------------------------------------
01 [5 5]              [14.  4.]          33.0000  [0.8 3.8]         
02 [0.8 3.8]          [5.6 1.6]          -11.5200  [-0.88  3.32]     
03 [-0.88  3.32]      [2.24 0.64]        -18.6432  [-1.552  3.128]   
04 [-1.552  3.128]    [0.896 0.256]      -19.7829  [-1.8208  3.0512] 
05 [-1.8208  3.0512]  [0.3584 0.1024]    -19.9653  [-1.9283  3.0205] 
06 [-1.9283  3.0205]  [0.1434 0.041 ]    -19.9944  [-1.9713  3.0082] 
07 [-1.9713  3.0082]  [0.0573 0.0164]    -19.9991  [-1.9885  3.0033] 
08 [-1.9885  3.0033]  [0.0229 0.0066]    -19.9999  [-1.9954  3.0013] 
09 [-1.9954  3.0013]  [0.0092 0.0026]    -20.0000  [-1.9982  3.0005] 
10 [-1.9982  3.0005]  [0.0037 0.001 ]    -20.0000  [-1.9993  3.0002] 
11 [-1.9993  3.0002]  [0.0015 0.0004]    -20.0000  [-1.9997  3.0001] 
12 [-1.9997  3.0001]  [0.0006 0.0002]    -20.0000  [-1.9999  3.    ] 
13 [-1.9999  3.    ]  [0.0002

In [32]:
print("ANSWER#1 min(J(w)): ", J(w50th))
print("ANSWER#2 w1: ", w50th[0])
print("ANSWER#3 w2: ", w50th[1])

ANSWER#1 min(J(w)):  -20.0
ANSWER#2 w1:  -2.0
ANSWER#3 w2:  3.0


In [33]:
w5th = wUpdatedIterate(np.array([5, 5]), 0.3, 4) # first iteration done, do rest 4.
print("------------")
print("ANSWER#4 w1: ", w5th[0])
print("ANSWER#5 w2: ", w5th[1])

k  w                  gradient J         J(w)        new w             
-----------------------------------------------------------
01 [5 5]              [14.  4.]          33.0000  [0.8 3.8]         
02 [0.8 3.8]          [5.6 1.6]          -11.5200  [-0.88  3.32]     
03 [-0.88  3.32]      [2.24 0.64]        -18.6432  [-1.552  3.128]   
04 [-1.552  3.128]    [0.896 0.256]      -19.7829  [-1.8208  3.0512] 
------------
ANSWER#4 w1:  -1.8208
ANSWER#5 w2:  3.0512


In [34]:
w4th = wUpdatedIterate(np.array([5, 5]), 0.3, 3)
w5th = wUpdatedIterate(np.array([5, 5]), 0.3, 4)
print("------------")
print("ANSWER#6 J5th(w)-J4th(w) :", round(abs(J(w5th)-J(w4th)), 4))

k  w                  gradient J         J(w)        new w             
-----------------------------------------------------------
01 [5 5]              [14.  4.]          33.0000  [0.8 3.8]         
02 [0.8 3.8]          [5.6 1.6]          -11.5200  [-0.88  3.32]     
03 [-0.88  3.32]      [2.24 0.64]        -18.6432  [-1.552  3.128]   
k  w                  gradient J         J(w)        new w             
-----------------------------------------------------------
01 [5 5]              [14.  4.]          33.0000  [0.8 3.8]         
02 [0.8 3.8]          [5.6 1.6]          -11.5200  [-0.88  3.32]     
03 [-0.88  3.32]      [2.24 0.64]        -18.6432  [-1.552  3.128]   
04 [-1.552  3.128]    [0.896 0.256]      -19.7829  [-1.8208  3.0512] 
------------
ANSWER#6 J5th(w)-J4th(w) : 0.1824


---------------
- The accepted answer is **0.1823**
- The above code gets **0.18235** and rounded up to  **0.1824**.
- Most common method of rounding rule is: round down if the next digit is less than 5. Round up otherwise.
---------------

#### Questions 7 to 10

In [0]:
import numpy as np
from numpy import linalg as LA
x = np.array([
    [1, 0],
    [1, 0.25],
    [1, 0.5],
    [1, 0.75],
    [1, 1.00]
]) # adding 1 in every inputs in order to accomodate bias/intercept.

y = np.array([
    [0.8822],
    [1.2165],
    [1.3171],
    [1.7930],
    [1.9826]
])

In [0]:
def J2(w):
    return 1.0/10*sum([(y[i][0]-w[1]*x[i][1]-w[0])**2 for i in range(5)])
                  
def gradientJ2(w):
    return np.array([1.0/10*sum([2*(y[i][0]-w[1]*x[i][1]-w[0])*-1 for i in range(5)]), # for w1
                     1.0/10*sum([2*(y[i][0]-w[1]*x[i][1]-w[0])*-x[i][1] for i in range(5)]) # for w2
                    ])

# Do the iteration specified in the iterate-parameter  
# Return the updated w 
# Note: If you want to get J(w) of fifth itertion, use the updated w from fourth iteration.  
def wUpdatedIterate2(wCurrent, alpha, iterate):
  
  print("%-2s %-18s %-18s %-12s %-18s" 
      % ("k", "w", "gradient J2", "J2(w)", "new w"))
  print("-----------------------------------------------------------")

  w = wCurrent
  for k in range(1, iterate+1):
    print("%02d %-18s %-18s %8.4f " 
          % (k, np.round(w,4), np.round(gradientJ2(w),4), J2(w)), end="")
    w = w - alpha * gradientJ2(w)   
    print(" %-16s" % (np.round(w,4)))
      
  return w

In [37]:
w1000th = wUpdatedIterate2(np.array([0, 0]), .3, 999) # first iteration done, do rest 999.

k  w                  gradient J2        J2(w)        new w             
-----------------------------------------------------------
01 [0 0]              [-1.4383 -0.858 ]    1.1138  [0.4315 0.2574] 
02 [0.4315 0.2574]    [-0.8781 -0.5457]    0.4334  [0.6949 0.4211] 
03 [0.6949 0.4211]    [-0.5328 -0.3526]    0.1741  [0.8548 0.5269] 
04 [0.8548 0.5269]    [-0.3201 -0.233 ]    0.0749  [0.9508 0.5968] 
05 [0.9508 0.5968]    [-0.1891 -0.1588]    0.0368  [1.0075 0.6445] 
06 [1.0075 0.6445]    [-0.1085 -0.1126]    0.0219  [1.0401 0.6782] 
07 [1.0401 0.6782]    [-0.0591 -0.0836]    0.0158  [1.0578 0.7033] 
08 [1.0578 0.7033]    [-0.0288 -0.0654]    0.0132  [1.0664 0.7229] 
09 [1.0664 0.7229]    [-0.0104 -0.0537]    0.0118  [1.0696 0.739 ] 
10 [1.0696 0.739 ]    [ 0.0008 -0.0461]    0.0110  [1.0693 0.7529] 
11 [1.0693 0.7529]    [ 0.0075 -0.041 ]    0.0104  [1.0671 0.7652] 
12 [1.0671 0.7652]    [ 0.0114 -0.0375]    0.0099  [1.0637 0.7764] 
13 [1.0637 0.7764]    [ 0.0136 -0.035 ]    0.0095  

In [38]:
# normal equation
wAccordingToNormalEquation = LA.inv(x.T.dot(x)).dot(x.T).dot(y)

print("W according to normal equation: %s " % wAccordingToNormalEquation)

from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(x[:,1:],y) # Remove all 1's added for accomodating bias/intercept
wAccordingToLinearRegression = lr.intercept_,lr.coef_
print("W according to linear equation: %s %s" % wAccordingToLinearRegression)
print("W according to gradient descnt: %s" % w1000th)

W according to normal equation: [[0.88282]
 [1.11092]] 
W according to linear equation: [0.88282] [[1.11092]]
W according to gradient descnt: [0.88282 1.11092]


In [39]:
w5th = wUpdatedIterate2(np.array([0, 0]), 1, 4)
print("------------")
print("ANSWER#7 :", len(w5th))
print("ANSWER#8 w0 :", round(w5th[0], 4))
print("ANSWER#9 w1 :", round(w5th[1], 4)) 

k  w                  gradient J2        J2(w)        new w             
-----------------------------------------------------------
01 [0 0]              [-1.4383 -0.858 ]    1.1138  [1.4383 0.858 ] 
02 [1.4383 0.858 ]    [0.429  0.1829]      0.0984  [1.0093 0.6751] 
03 [1.0093 0.6751]    [-0.0914 -0.1002]    0.0184  [1.1007 0.7753] 
04 [1.1007 0.7753]    [ 0.0501 -0.0169]    0.0107  [1.0506 0.7922] 
------------
ANSWER#7 : 2
ANSWER#8 w0 : 1.0506
ANSWER#9 w1 : 0.7922


In [40]:
w5th = wUpdatedIterate2(np.array([0, 0]), 1, 4)
print("ANSWER10 J5th(w) :", round(J2(w5th), 4))

k  w                  gradient J2        J2(w)        new w             
-----------------------------------------------------------
01 [0 0]              [-1.4383 -0.858 ]    1.1138  [1.4383 0.858 ] 
02 [1.4383 0.858 ]    [0.429  0.1829]      0.0984  [1.0093 0.6751] 
03 [1.0093 0.6751]    [-0.0914 -0.1002]    0.0184  [1.1007 0.7753] 
04 [1.1007 0.7753]    [ 0.0501 -0.0169]    0.0107  [1.0506 0.7922] 
ANSWER10 J5th(w) : 0.0088


In [41]:
w6th = wUpdatedIterate2(np.array([0, 0]), 1, 6)

k  w                  gradient J2        J2(w)        new w             
-----------------------------------------------------------
01 [0 0]              [-1.4383 -0.858 ]    1.1138  [1.4383 0.858 ] 
02 [1.4383 0.858 ]    [0.429  0.1829]      0.0984  [1.0093 0.6751] 
03 [1.0093 0.6751]    [-0.0914 -0.1002]    0.0184  [1.1007 0.7753] 
04 [1.1007 0.7753]    [ 0.0501 -0.0169]    0.0107  [1.0506 0.7922] 
05 [1.0506 0.7922]    [ 0.0085 -0.0356]    0.0088  [1.0422 0.8278] 
06 [1.0422 0.8278]    [ 0.0178 -0.0265]    0.0076  [1.0244 0.8543] 
